A52-xds-custom-lb-policies.md

```diff
@@ -0,0 +1,312 @@
+}
+
+message LoadBalancingConfig {
+  oneof policy {
+
+    …
+
+    optional XdsWrrLocalityLoadBalancingPolicyConfig xds_wrr_locality_experimental = 16
+      [json_name = "xds_wrr_locality_experimental"];
+  }
+```
+
+This load balancer will be responsible for generating the configuration of its child policy, `weighted_target_experimental`. To do this, two sources of information will be combined:
+
+- The newly defined xDS WRR Locality configuration with the child policy config
+- Locality weights from a new resolved address attribute
+  - Passed in similar manner as the xDS client
+  - Populated by the `xds_cluster_resolver` policy
+
+These two elements will be combined to create the `weighted_target_experimental` configuration with a weighted policy selection for each locale. The format of this configuration will not change, allowing the `weighted_target_experimental` policy to remain unchanged.
+
+![gRPC Client Architecture Diagram](A52_graphics/grpc_xds_client_architecture.png)
+
+[Link to SVG file](A52_graphics/grpc_xds_client_architecture.svg)
+
+### xDS Cluster Resolver Load Balancer
+
+As described in [gRFC A42](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md#change-child-policy-config-generation-in-xds_cluster_resolver-policy), this load balancer today determines the load balancer hierarchy when it creates the configuration for the priority policy. If the policy it receives is `ring_hash`, it will pass that policy selection directly to the priority policy. If it is given an endpoint picking policy (like `round_robin`), it will instead create a weighted target policy selection that contains the endpoint picking policy selection as a child, creating a two level hierarchy.
+
+With the xDS client now creating a configuration that represents the whole policy hierarchy, this special logic can be removed and the policy selection received can be directly passed to the priority policy.
+
+The cluster resolver configuration documentation in service_config.proto is updated to reflect this:
+
+```protobuf
+// Configuration for xds_cluster_resolver LB policy.
+message XdsClusterResolverLoadBalancingPolicyConfig {
+  // xDS LB policy. Will be used as the child config of the xds_cluster_impl LB policy. 
+  repeated LoadBalancingConfig xds_lb_policy = 2;
+
+}
+```
+
+To provide the `xds_wrr_locality` load balancer information about locality weights received from EDS, the cluster resolver will populate a new attribute in the resolved addresses. The new attribute will contain a map from a locality struct to a locality weight integer. Note that a change in this attribute would still allow associated subchannels to be reused - it will not affect their uniqueness.
+
+### Configuration Example
+
+This is a step-by-step example of how and example xDS configuration gets converted within a gRPC client. The example configuration is:
+
+```textproto
+{ // Cluster
+  load_balancing_policy: {
+    policies: [
```

Comment:
`[` for messages doesn't look like textproto format. The syntax highlighter is also complaining about this.

Reply:
Textproto does actually support `[...]` for repeated fields these days.

Reply:
@markdroth, it supports `[...]` it for repeated messages? I've only seen it for repeated scalars. (In text_format.h, I think that is `SetUseShortRepeatedPrimitives(true)`.)